File: app/types/will.ts
import { PublicKey } from "@solana/web3.js";

export interface Will {
  testator: PublicKey;
  beneficiary: PublicKey;
  vault: PublicKey;
  heartbeatPeriod: number;
  status: WillStatus;
  createdAt: number;
  lastHeartbeat: number;
  triggerAt: number | null;
  bump: number;
  vaultBump: number;
}

export enum WillStatus {
  Created = 0,
  Active = 1,
  Triggered = 2,
  Claimed = 3,
  Withdrawn = 4,
}

export interface GlobalConfig {
  authority: PublicKey;
  feeVault: PublicKey;
  fee: number;
  tokenFeeBps: number;
  nftFeeLamports: number;
  minHeartbeatPeriod: number;
  maxHeartbeatPeriod: number;
  paused: boolean;
  bump: number;
}

export interface WillFormData {
  beneficiary: string;
  heartbeatPeriod: number;
}

export interface DepositFormData {
  amount: number;
}

File: backend/pkg/types/will.go
package types

import "github.com/gagliardetto/solana-go"

type WillStatus uint8

const (
	WillStatusCreated   WillStatus = 0
	WillStatusActive    WillStatus = 1
	WillStatusTriggered WillStatus = 2
	WillStatusClaimed   WillStatus = 3
	WillStatusWithdrawn WillStatus = 4
)

type Will struct {
	Testator        solana.PublicKey
	Beneficiary     solana.PublicKey
	Vault           solana.PublicKey
	HeartbeatPeriod uint32
	Status          WillStatus
	CreatedAt       int64
	LastHeartbeat   int64
	TriggerAt       *int64
	Bump            uint8
	VaultBump       uint8
}